app/javascript/packs/document-capture.jsx

```diff
@@ -38,6 +38,9 @@ loadPolyfills(['fetch']).then(() => {
         endpoint={appRoot.getAttribute('data-endpoint')}
         csrf={getMetaContent('csrf-token')}
         isMockClient={isMockClient}
+        errorRedirects={{
+          429: appRoot.getAttribute('data-throttled-url'),
+        }}
```

Comment:
This seems weird to me, I think we should catch the response from the vendor and handle it better in our API controller

What do you think of:
1. Just returning a more clear status and just having the clients render the error rather than redirect?
2. Or, we could update the error response to include the URL to redirect to? Rather than this separate "global" config?

I think it would map better for when we do async thing and have to poll, because checking the response code of a endpoint that we poll seems kind of unclear?

Reply:
With the current implementation, it looks like you could only have one URL to redirect to? I'm not sure if we need more than one right now, but I think either 1 or 2 are good solutions to handling that.

Reply:
To the first point, I'd [raised this in Slack](https://gsa-tts.slack.com/archives/CNCGEHG1G/p1599837573144100) as potentially being something where we could just use the existing error handling, though it was my understanding that since the screen exists already, the redirect behavior was considered the simplest approach.

To the idea of including it with the error response, I can see how that might work a little better. With how it's currently implemented, I could imagine it either being a property of an individual error (`errors[0].redirect`), or since we're already considering "too many requests" as the canonical error (represented by the status code 429), maybe it's a response header (`X-Redirect`)? It could also be that we respond with a 302 redirect, though it doesn't seem ideal from my perspective, since we'd be losing information that the request was rejected. Not sure if it's a horrible abuse, but maybe the standard [`Location` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location) would be fine to use in combination with the `429` status code?

Reply:
I like the idea of an explicit `redirect` property better than an HTTP header, I think if we're making a JSON API, the contents of the response should be where the relevant data lives

```js
{
  errors: [
    { redirect: '/429' }
  }
}
```

```js
let redirectError = errors.find(e => e.redirect);
if (redirectError) {
  // redirect
} else {
  render(errors)
}
```

Reply:
I guess my point was more that we already prior art with handling errors in aggregate, as in how we assign the status code of the response:

https://github.com/18F/identity-idp/blob/0ebeb69bcd667f6595fad0b0a3307b2adb6413a7/app/forms/idv/api_image_upload_form.rb#L37-L41

So if it wouldn't be too out of place to then handle the redirect as a response header.

But I don't feel too strongly one way or the other. I think it'd be reasonable enough to include it as a property of the individual error. 👍

Reply:
Wow I already forgot about that. I'd say all "bad" statuses would be just plain 400, then we can remove/clean up that status logic if we make things part of the errors array

Reply:
So, I'm still on the fence here, and in exploring this further found additional complications.

The most compelling argument for me in favor of your recommendation is the mismatch that we have where there are potentially many individual errors, yet we still choose an error code 429 to return, associated with only one of these errors.

On the other hand, I think 429 is the sensible status code to return for a rate-limited user when we know that's the error. A client should be capable to detect this status code and deal with it appropriately. For example, a generic client might receive a 429 and back off for a few minutes. In our case, that handling is the redirect. I think it's made a little muddy by the fact that our handling is redirecting to a page specific to document capture throttling. But maybe this isn't too problematic, because both the page/partial and the JavaScript pack itself are intending to implement document capture.

As for complications: Getting additional details (`redirect` property) to the error response is a little tricky, since the standard assumed type of `FormResponse.errors` is an array of string messages (see also [`first_error_message`](https://github.com/18F/identity-idp/blob/0ebeb69bcd667f6595fad0b0a3307b2adb6413a7/app/services/form_response.rb#L26-L30)). My original hunch was to take advantage of `ActiveModel::Errors.details`, and [`add`](https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-add) an error like `errors.add(:limit, t('errors.doc_auth.acuant_throttle'), redirect: idv_session_errors_throttled_url)`. But there's not a direct way to get these details from [`ApiImageUploadForm`](https://github.com/18F/identity-idp/blob/master/app/forms/idv/api_image_upload_form.rb) to [`ImageUploadResponsePresenter`](https://github.com/18F/identity-idp/blob/0ebeb69bcd667f6595fad0b0a3307b2adb6413a7/app/presenters/image_upload_response_presenter.rb) without misusing the `FormResponse.errors` as either the details array, or as the instance of `ActiveModel::Errors`. The most reasonable non-misused approach I considered is to shove the extra field details into `FormResponse.extra`.

As for how to move forward, I'm now wondering about a totally different option, specifically to the earlier point that there's a mismatch between the status code and individual errors. Instead, what if we still return a 429, but treat it as a singular error, and not bother with the `errors` array (or not bother with a JSON response at all). I think this could be reasnonble since we're not expecting to use the errors in the response anyways, or there'd only ever be the one `:limit` error.

So we'd have:

- Rate limited
   - Status code: `429`
   - Body: _(empty)_ (or alternatively: `{"success": false, "redirect": "..."}`)
   - Headers: `X-Redirect: ...` (or absent, if following above "alternatively")
- Any other errors
   - Status code: `400`
   - Body: `{"success": false, "errors": [ ... ]}`

Reply:
Yeah this makes sense, having a different top-level keys in the response, ex `redirect` vs `errors`

Reply:
> Yeah this makes sense, having a different top-level keys in the response, ex `redirect` vs `errors`

👍  Updated in ea1dedc.